### PR TITLE
Added the amphibious mode to the Tiberian Sun APC

### PR DESCRIPTION
**Pull request by Mailaender**
image

### CODE REVIEW

File: OpenRA.Mods.TS/Traits/Render/WithVoxelWaterBody.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+	{
+		public readonly string WaterSequence = "water";
+		public readonly string LandSequence = "idle";
+
+		public object Create(ActorInitializer init) { return new WithVoxelWaterBody(init.Self, this); }
+
+		public IEnumerable<VoxelAnimation> RenderPreviewVoxels(ActorPreviewInitializer init, RenderVoxelsInfo rv, string image, WRot orientation, int facings, PaletteReference p)
+		{
+			var body = init.Actor.Traits.Get<BodyOrientationInfo>();
+			var voxel = VoxelProvider.GetVoxel(image, LandSequence);
```
Comment (pchote) [inline OpenRA.Mods.TS/Traits/Render/WithVoxelWaterBody.cs]: You can query the location from the init and then use the WaterSequence when it is placed on water.

File: OpenRA.Mods.TS/Traits/Render/WithVoxelWaterBody.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+				() => new[] { body.QuantizeOrientation(orientation, facings) },
+				() => false, () => 0);
+		}
+
+		public int QuantizedBodyFacings(ActorInfo ai, SequenceProvider sequenceProvider, string race) { return 0; }
+	}
+
+	public class WithVoxelWaterBody : IAutoSelectionSize
+	{
+		bool Water { get { return self.World.Map.GetTerrainInfo(self.Location).IsWater; } }
```
Comment (pchote) [inline OpenRA.Mods.TS/Traits/Render/WithVoxelWaterBody.cs]: Our naming conventions for bools are <preposition><noun>, so I suggest calling this `OverWater`.  It should also live below the readonly fields.

**Comment by pchote** [comment]
Could you make this depend on #7974 (or vice versa) and then implement the water wake?

File: OpenRA.Mods.TS/Traits/Render/WithVoxelWaterBody.cs (hunk 2 above)
Comment (penev92) [inline OpenRA.Mods.TS/Traits/Render/WithVoxelWaterBody.cs]: With `Is` in front for bools, please.

**Comment by Mailaender** [comment]
Updated.

**Comment by penev92** [comment]
I'm not going to request this, as it it quite outside the scope of this implementation. I just want to note that I would have liked to see this done with something like `UpgradeOnTerrain`, which can then be used for other things as well (like replacing `TerrainModifiesDamage` with a proper upgrade).

**Comment by epicelite** [comment]
TS when?

**Comment by pchote** [comment]
@epicelite: We reset the countdown to 7 months each time somebody asks that question.

**Comment by pchote** [comment]
Needs a rebase now that the dependencies have been merged.

**Comment by Mailaender** [comment]
Rebased.

File: mods/ts/rules/gdi-vehicles.yaml [hunk 3]
```diff
@@ -12,6 +12,8 @@ APC:
 	Mobile:
 		ROT: 5
 		Speed: 113
+		TerrainSpeeds:
+			Water: 70
```
Comment (penev92) [inline mods/ts/rules/gdi-vehicles.yaml]: Can someone with a working knowledge of Tiberian Sun unit stats (or with TibEd and Tiberian Sun installed) confirm that this value is correct please?
(@reaperrr, @abcdefg30)

Comment (reaperrr) [inline mods/ts/rules/gdi-vehicles.yaml]: Should be 80 according to TS' rules.ini.

**Comment by penev92** [comment]
I'm going to bet this is caused by faulty tile definitions, as I'm pretty sure this isn't supposed to be classified as `TerrainType: Water`:


I also noticed some visual glitches happening around the APC while it is turning. Most noticeable when turning from facing Sount-West to North-West (again I'm sure it has nothing to do with this PR, but it can turn out to be serious).

P.S.: Also lol:

**Comment by penev92** [comment]
Other than the unrelated issues mentioned above (which **need** to be addressed!), this seems to work well.
:+1: when the movement speed is fixed.

**Comment by abcdefg30** [comment]
See also https://github.com/OpenRA/OpenRA/pull/7974#issuecomment-94255920.

File: OpenRA.Mods.TS/Traits/Render/WithVoxelWaterBody.cs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+			size = new int2(s, s);
+
+			var waterVoxel = VoxelProvider.GetVoxel(rv.Image, info.WaterSequence);
+			rv.Add(new VoxelAnimation(waterVoxel, () => WVec.Zero,
+				() => new[] { body.QuantizeOrientation(self, self.Orientation) },
+				() => !OverWater,
+				() => 0));
+		}
+
+		public int2 SelectionSize(Actor self) { return size; }
```
Comment (Phrohdoh) [inline OpenRA.Mods.TS/Traits/Render/WithVoxelWaterBody.cs]: `self` isn't necessary here.

File: OpenRA.Mods.TS/Traits/Render/WithVoxelWaterBody.cs [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+				() => false, () => 0);
+		}
+
+		public int QuantizedBodyFacings(ActorInfo ai, SequenceProvider sequenceProvider, string race) { return 0; }
+	}
+
+	public class WithVoxelWaterBody : IAutoSelectionSize
+	{
+		readonly Actor self;
+		readonly int2 size;
```
Comment (Phrohdoh) [inline OpenRA.Mods.TS/Traits/Render/WithVoxelWaterBody.cs]: Why not make this public instead of wrapping it in a method?

**Comment by Phrohdoh** [comment]
I'd like my comments to be addressed but it isn't absolutely necessary.
:+1: ready for merge once the speed is fixed.

Works fine ingame, thanks. :-)

**Comment by Mailaender** [comment]
I was also wondering about that. Thought it was meant to be ice.

**Comment by penev92** [comment]
Yeah it looks like we have a bug in the importer (@pchote).

**Comment by pchote** [comment]
The importer is just responding to the data provided by the original game files and @tomsons26.

File: mods/ts/rules/gdi-vehicles.yaml (hunk 3 above)
Comment (Mailaender) [inline mods/ts/rules/gdi-vehicles.yaml]: Updated.

File: OpenRA.Mods.TS/Traits/Render/WithVoxelWaterBody.cs (hunk 4 above)
Comment (Mailaender) [inline OpenRA.Mods.TS/Traits/Render/WithVoxelWaterBody.cs]: That is required to implement the 
```csharp
public interface IAutoSelectionSize { int2 SelectionSize(Actor self); }
```

File: OpenRA.Mods.TS/Traits/Render/WithVoxelWaterBody.cs (hunk 5 above)
Comment (Mailaender) [inline OpenRA.Mods.TS/Traits/Render/WithVoxelWaterBody.cs]: I won't change this convention here. My guess this is required for `RenderSprites.AutoSelectionSize` which is a bit more complicated due to all the overlays factor into the selection size calculation.

**Comment by Mailaender** [comment]
> I'd like my comments to be addressed but it isn't absolutely necessary.

Your comments are mostly unrelated to this pull request so I can't address them here, but I also did not ignore them. Hope this is okay.

**Comment by Mailaender** [comment]
> P.P.S.: There seems to be a 3x4 pathable area on the river, as shown [here](http://i.imgur.com/lJh2z7G.png).

Tiberian Sun has an ice growing, retracting and cracking mechanic that we also track at https://github.com/OpenRA/OpenRA/issues/7874. I believe these tiles are used for that feature dynamic, but we have not implemented the logic required for it.

File: OpenRA.Mods.TS/Traits/Render/WithVoxelWaterBody.cs (hunk 4 above)
Comment (Phrohdoh) [inline OpenRA.Mods.TS/Traits/Render/WithVoxelWaterBody.cs]: Okay that makes sense. XS didn't show me that for some reason.

**Pull request merged**